### DISCUSSION TITLE
Dict key autocomplete not working when importing the dictionnary

### DISCUSSION THREAD

**Discussion by ClementJ18** [microsoft/pylance-release#4771] [Q&A] [answered]
This is a minimal example:

stuff.py
```py
the_dict = {
    "Goblin": 1,
    "Bigger Goblin": 2
}

the_dict[""]
```

main.py
```py
from stuff import the_dict

the_dict[""]
```

If you start typing a key name in foo.py it will autocomplete as expected from a static dict
![image](https://github.com/microsoft/pylance-release/assets/22744199/51a8a9f5-1041-4068-8c1c-46e9ab551b94)

However, if you import the dict into main.py the autocomplete will fail
![image](https://github.com/microsoft/pylance-release/assets/22744199/314746d5-9340-42ab-9fde-2c706d761032)

Is there a way to get the autocomplete working? Or is this simply not possible? In my use case I'm using this dict to store all enemies for a game which makes it quite large (or will once I'm done adding them). I wanted to find a data structure that would allow for auto-complete first.

**Comment by rchiodo** [comment]
Seems like a bug but there is a workaround. If you create the dict as a TypedDict then the completions will work elsewhere:

```python
from typing import TypedDict


class Enemies(TypedDict):
    Goblin: int
    Hobgoblin: int

the_dict: Enemies = {
    "Goblin": 1,
    "Hobgoblin": 2
}
```

Usage:

![image](https://github.com/microsoft/pylance-release/assets/19672699/5efef9f1-c9bb-4241-aed9-921d300e2b1f)

**Reply by erictraut** [reply, marked as answer]
I think this is working as intended. The completion provider (the component that presents completion suggestions) bases its suggestions on type information. With an untyped dict such as `the_dict` in the first example, the inferred type is simply `dict[str, int]`. That type doesn't contain any information about the keys in the dict. And since it's a mutable type, the keys could be changed dynamically.

IIRC, there was some special-case logic added to pylance's completion provider for locally-assigned dict expressions that allow it to offer key name suggestions from the dict expression within the module. This explains why this works locally but doesn't work across module boundaries. 

Defining a TypedDict requires a bit of extra work, but it provides sufficient information for the completion provider to suggest detailed key information even across module boundaries. It has the added advantage that the type checker will now enforce the type of each key in the dictionary — something that's really useful if you have a heterogeneous dictionary.

**Reply by ClementJ18** [reply]
Thanks for your replies. Creating a TypedDict would create a lot of duplication. I will probably create a "Beastiary" class and attach all the enemies to it as class attributes then. Something like
```py
class Enemies
    Goblin = 1
    HobGoblin = 2
```
